src/phoenix/core/traces.py

```diff
@@ -163,12 +163,16 @@ def get_spans(
             if root_spans_only
             else self._start_time_sorted_span_ids
         )
-        for span_id in sorted_span_ids.irange_key(
-            start_time.astimezone(timezone.utc),
-            stop_time.astimezone(timezone.utc),
-            inclusive=(True, False),
-            reverse=True,  # most recent spans first
-        ):
+        with self._lock:
```

Comment:
question: feels like it's up to the programmer to know to grab the lock - is there a way this can be done implicitly so it's not up to the programmer?

Reply:
Yea, mutex is one of the areas where async really shines: with async the surrender of control is voluntary so no locks are necessary. Ultimately, if our application becomes more high-stakes than it currently is, we would have to rethink concurrency more thoroughly. But looking closely, most of the locks that I'm adding now are probably not necessary, because the standard Python interpreter is single-threaded, which means Python itself behaves like an event loop, but with involuntary preemptions and has a default switch interval of 5ms. Most of our operations probably won't take 5ms anyway, so I was ambivalent about adding the locks. However, I became somewhat concerned about the third party libraries that we are using (now or in the future). So I just started adding locks to everything, in part to address your exact question about when to use locks: let's just use them all the time. Each lock costs maybe a hundred nanoseconds, and we are nowhere near where contention is an issue.